### PR TITLE
Book 11 Cron Jobs & Bull Jobs & Set up everything together

### PR DESCRIPTION
**Pull request by Getriax**
## Description

The bot should send a message 10 minutes before the event, asking for a confirmation. If 5 minutes before the start the event is not confirmed the rooms should be removed.

## Changes

* Implemented bull module
* added cron jobs
* created the `controller`
* created the `services` folder that uses previously created ones
* created `events.dao` for managing the data exchange between bot and Redis

### CODE REVIEW

File: src/clients/GoogleClientExtractorFactory.js [hunk 1]
```diff
@@ -83,19 +84,23 @@ class GoogleClient {
     if (token) {
       ({ access_token: accessToken } = token);
     }
-
+    console.log(accessToken);
```
Comment (kozikkam) [inline src/clients/GoogleClientExtractorFactory.js]: What's the point of logging access token?

File: src/clients/GoogleClientExtractorFactory.js [hunk 2]
```diff
@@ -118,9 +123,12 @@ class GoogleClient {
...
-        this.oAuthClient.getToken(code, this.setCredentials);
-        this.messageMasterRoom('Authorization succeed!');
-
+        try {
+          this.oAuthClient.getToken(code, this.setCredentials);
+          this.messageMasterRoom('Authorization succeeded');
+        } catch (error) {
+          return res.status(400).send('Could not authorize');
+        }
         return res.status(200).send('You can close this window now, thanks!');
```
Comment (kozikkam) [inline src/clients/GoogleClientExtractorFactory.js]: Might want an empty line before return.

File: test/book/book.processes.spec.js [hunk 3]
```diff
@@ -0,0 +1,184 @@
+/* eslint-disable no-underscore-dangle */
+const Helper = require('hubot-test-helper');
+const sinon = require('sinon');
+const chai = require('chai');
+
+const { expect } = chai;
```
Comment (kozikkam) [inline test/book/book.processes.spec.js]: Might as well just const { expect } = require('chai');

File: src/clients/createRedis.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      ownerEmail: faker.internet.email(),
+      name: faker.lorem.words(2),
+      url: faker.internet.url(),
+      confirmed,
+      requested: confirmed ? true : faker.random.boolean(),
+    }),
+  };
+};
+
+// const promises = [];
```
Comment (krzyurb) [inline src/clients/createRedis.js]: Please remove commented code. And why this file is in `/src/clients/`? It looks like factory for tests

File: src/clients/createRedis.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// for (let i = 0; i < 30; i += 1) {
+//   const { key, value } = randomEvent();
+//   promises.push(redis.hset(main, key, value).then(() => console.log('Added', { key }, '\n', { value: JSON.parse(value) })));
+// }
+
+// Promise.all(promises)
+//   .then(() => process.exit(0))
+//   .catch(console.error);
+
+redis.hsetnx('keys', 'bbb', 'BBB')
```
Comment (krzyurb) [inline src/clients/createRedis.js]: What is the purpose of that?

File: test/utils/QuqueJob.spec.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
+const chai = require('chai');
+const sinon = require('sinon');
+const faker = require('faker');
+
+const { expect } = chai;
+const { QueueJob } = require('../../utils');
+
+const job = (delay) => {
```
Comment (krzyurb) [inline test/utils/QuqueJob.spec.js]: How about moving it to `factories` directory?

File: test/utils/QuqueJob.spec.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+describe('QueueJob test suite', () => {
+  let queueJob;
+
+  beforeEach(() => {
+    sinon.stub(QueueJob, 'crateQueue').returns(client);
+    queueJob = new QueueJob('job');
+  });
+
+  afterEach(() => sinon.restore());
+
```
Comment (krzyurb) [inline test/utils/QuqueJob.spec.js]: Remove empty line

File: utils/QueueJob/src/QueueJob.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
+const Queue = require('bull');
+
+const { REDIS_URL } = process.env;
```
Comment (krzyurb) [inline utils/QueueJob/src/QueueJob.js]: Please add info about `REDIS_URL` to plugin documentation

File: scripts/book.js [hunk 9]
```diff
@@ -4,16 +4,37 @@
...
   dependenciesLocator.set('slackActionsMultiplexer', new SlackActionsMultiplexer(robot));
 
   const slackActionsMultiplexer = dependenciesLocator.get('slackActionsMultiplexer');
   slackActionsMultiplexer.addAction(/confirm-booking/, actions.confirmBooking);
+
+  const googleClientExtractor = await googleClientExtractorFactory(robot);
+
+  const bookService = new BookService(
+    robot, googleClientExtractor, slackClient, eventsDAO,
+  );
```
Comment (krzyurb) [inline scripts/book.js]: Please add empty line below

**Comment by Getriax** [comment]
@krzyurb & @kozikkam 🛠 done

**Pull request merged**